scripts/pull-rhoc-balances/balances.js

```diff
@@ -0,0 +1,99 @@
+/// balances.js -- RHOC balance reporting
+
+// Copyright (C) 2018 dc <dc@dapp.org>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+const R    = require('ramda');
+const fs   = require('fs');
+const Web3 = require('web3');
+const bb   = require('bluebird');
+
+const web3 = new Web3(process.env.ETH_WS || 'wss://mainnet.infura.io/_ws');
+const rhoc = new web3.eth.Contract(require('./abi.json'), "0x168296bb09e24a88805cb9c33356536b980d3fc5");
+const keys = new Set();
+
+const fromBlock = 3383352;
+const toBlock   = process.env.BLOCK || 5866762;
+const batch     = 500
+
+// Block range increments
+const blockRange = () => R.map(
+  n => fromBlock + batch * n,
+  R.range(0, (1 + (toBlock - fromBlock) / batch) >>> 0)
+);
+
+const blockPairs = R.aperture(2, blockRange());
+
+const writeStream = fs.createWriteStream('balances.csv');
+
+// Parse all transfer events between a pair of blocks
+const parseTransfers = (pair) => {
+  return rhoc.getPastEvents('Transfer', { fromBlock: pair[0], toBlock: pair[1] })
+    .then(transfers => {
+      bb.map(transfers, (t) => {
+        return [t.returnValues.from, t.returnValues.to].forEach(addKey)
+      }, {concurrency: 5})
+    })
+    .catch(e => error(e, pair));
+}
+
+// Only write unique keys
+const addKey = (key) => { if (!keys.has(key)) { keys.add(key); write(key); } }
+
+// Retrieve key state and append to file
+const write = (key) => {
+  return rhocBalance(key).then(bal => {
+    if (bal > 0) {
+      return isContract(key).then(c => {
```

Comment:
This script checks for all RHOC balances at contracts. What we probably want is a script that checks for all RHOC balances at all addresses regardless of contract or not.

Reply:
As @TomVasile pointed out, this just passes through whether or not it is a contract. Ignore the above.